File: src/waypaper/wallhaven.py
"""Wallhaven API client — fetch wallpapers, thumbnails, and full-res downloads"""

import json
import time
import urllib.request
import urllib.parse
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

API = "https://wallhaven.cc/api/v1"
UA = "waypaper-wallhaven/1.0"

CAT_PRESETS = {
    "random":   {"cats": "111", "purity": "100", "sorting": "random"},
    "anime":    {"cats": "010", "purity": "100", "sorting": "random"},
    "manga":    {"cats": "010", "purity": "100", "q": "manga panel", "sorting": "random"},
    "sketch":   {"cats": "111", "purity": "010", "sorting": "random"},
    "general":  {"cats": "100", "purity": "100", "sorting": "random"},
}


@dataclass
class WallpaperItem:
    id: str
    thumb_url: str
    full_url: str
    resolution: str
    tags: list[str]


@dataclass
class PageMeta:
    current: int = 1
    last: int = 1
    total: int = 0


_last_req = 0.0


def _request(url: str) -> dict:
    global _last_req
    elapsed = time.time() - _last_req
    if elapsed < 1.5:
        time.sleep(1.5 - elapsed)
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    with urllib.request.urlopen(req, timeout=15) as resp:
        _last_req = time.time()
        return json.loads(resp.read().decode())


def search(preset: str = "random", query: str = "", page: int = 1,
           categories: str = None, purity: str = None, sorting: str = None,
           top_range: str = "", atleast: str = "",
           ai_art_filter: int = None, ratios: str = "", colors: str = "") -> tuple[list[WallpaperItem], PageMeta]:
    """Search Wallhaven with full parameter support.

    If categories/purity/sorting are provided as kwargs they override the preset.
    """
    params = {"page": str(page)}

    if categories is not None:
        params["categories"] = categories
        params["purity"] = purity if purity is not None else "100"
        params["sorting"] = sorting if sorting is not None else "date_added"
        if query:
            params["q"] = query
    else:
        p = CAT_PRESETS.get(preset, CAT_PRESETS["random"])
        params["categories"] = p["cats"]
        params["purity"] = p["purity"]
        params["sorting"] = p["sorting"]
        if query:
            params["q"] = query
        elif p.get("q"):
            params["q"] = p["q"]

    if top_range:
        params["topRange"] = top_range
    if atleast:
        params["atleast"] = atleast
    if ai_art_filter is not None:
        params["ai_art_filter"] = str(ai_art_filter)
    if ratios:
        params["ratios"] = ratios
    if colors:
        params["colors"] = colors

    url = f"{API}/search?{urllib.parse.urlencode(params)}"
    data = _request(url)

    meta = PageMeta(
        current=data.get("meta", {}).get("current_page", 1),
        last=data.get("meta", {}).get("last_page", 1),
        total=data.get("meta", {}).get("total", 0),
    )

    items = []
    for item in data.get("data", []):
        items.append(WallpaperItem(
            id=item.get("id", ""),
            thumb_url=item.get("thumbs", {}).get("small", ""),
            full_url=item.get("path", ""),
            resolution=item.get("resolution", ""),
            tags=[t.get("name", "") for t in item.get("tags", [])],
        ))

    return items, meta


def fetch_thumbnail(thumb_url: str):
    """Download a thumbnail from the CDN and return raw bytes.

    Returns a Pixbuf on Linux (gi), raw bytes on Windows.
    Callers should handle both: on the web path the bytes are sent directly.
    """
    if not thumb_url:
        return None
    try:
        req = urllib.request.Request(thumb_url, headers={"User-Agent": UA})
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = resp.read()

        # On Linux we return a GdkPixbuf; on Windows raw bytes.
        try:
            import gi
            gi.require_version("GdkPixbuf", "2.0")
            from gi.repository import GdkPixbuf
            loader = GdkPixbuf.PixbufLoader()
            loader.write(data)
            loader.close()
            return loader.get_pixbuf()
        except (ImportError, ValueError):
            return data
    except Exception:
        return None


def download_full(full_url: str, dest: Path) -> Optional[Path]:
    """Download full-resolution wallpaper to dest. Returns dest on success."""
    try:
        dest.parent.mkdir(parents=True, exist_ok=True)
        req = urllib.request.Request(full_url, headers={"User-Agent": UA})
        with urllib.request.urlopen(req, timeout=30) as resp:
            dest.write_bytes(resp.read())
        return dest
    except Exception:
        return None

File: electron-app/src/main/wallhaven.ts
export interface WallpaperItem {
  id: string
  thumb_url: string
  full_url: string
  resolution: string
  tags: TagInfo[]
  purity: string
  category: string
  file_size: number
  upload_date: string
  likes: number
  views: number
}

export interface TagInfo {
  name: string
  is_nsfw?: boolean
}

export interface PageMeta {
  current: number
  last: number
  total: number
}

export interface SearchParams {
  categories: string     // '100' (general) | '010' (anime) | '001' (people) | '111' (all)
  purity: string         // '100' (sfw) | '010' (sketchy) | '001' (nsfw) | '111' (all)
  sorting: string        // 'date_added' | 'relevancy' | 'random' | 'views' | 'favorites' | 'toplist'
  topRange?: string      // '1d' | '3d' | '1w' | '1m' | '3m' | '6m' | '1y'
  query?: string
  page: number
  atleast?: string       // minimum resolution e.g. '1920x1080'
  ratios?: string        // comma-separated, e.g. '16x9,16x10'
  colors?: string        // hex color without #
  ai_art_filter?: number // 0 = show AI, 1 = hide AI
}

const API = 'https://wallhaven.cc/api/v1'
const UA = 'waypaper-wallhaven/1.0'

let lastReq = 0
let requestCount = 0
const RATE_LIMIT_MS = 1500
const MAX_RETRIES = 2

async function request(url: string): Promise<any> {
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      const elapsed = Date.now() - lastReq
      if (elapsed < RATE_LIMIT_MS) {
        await new Promise(r => setTimeout(r, RATE_LIMIT_MS - elapsed))
      }
      const controller = new AbortController()
      const timeout = setTimeout(() => controller.abort(), 15000)
      const res = await fetch(url, {
        headers: { 'User-Agent': UA },
        signal: controller.signal,
      })
      clearTimeout(timeout)
      if (!res.ok) {
        if (res.status === 429 && attempt < MAX_RETRIES) {
          await new Promise(r => setTimeout(r, 3000))
          continue
        }
        throw new Error(`Wallhaven API error: ${res.status}`)
      }
      lastReq = Date.now()
      requestCount++
      return res.json()
    } catch (err: any) {
      if (attempt >= MAX_RETRIES) throw err
      if (err.name === 'AbortError') {
        await new Promise(r => setTimeout(r, 2000))
        continue
      }
      throw err
    }
  }
}

export async function search(params: SearchParams): Promise<{ items: WallpaperItem[]; meta: PageMeta }> {
  const urlParams = new URLSearchParams({
    categories: params.categories,
    purity: params.purity,
    sorting: params.sorting,
    page: String(params.page),
  })
  if (params.query) urlParams.set('q', params.query)
  if (params.topRange) urlParams.set('topRange', params.topRange)
  if (params.atleast) urlParams.set('atleast', params.atleast)
  if (params.ratios) urlParams.set('ratios', params.ratios)
  if (params.colors) urlParams.set('colors', params.colors)
  if (params.ai_art_filter !== undefined) urlParams.set('ai_art_filter', String(params.ai_art_filter))

  const url = `${API}/search?${urlParams.toString()}`
  const data = await request(url)

  const meta: PageMeta = {
    current: data.meta?.current_page ?? 1,
    last: data.meta?.last_page ?? 1,
    total: data.meta?.total ?? 0,
  }

  const items: WallpaperItem[] = (data.data || []).map((item: any) => ({
    id: item.id || '',
    thumb_url: item.thumbs?.small || '',
    full_url: item.path || '',
    resolution: item.resolution || '',
    tags: (item.tags || []).map((t: any) => ({ name: t.name || '', is_nsfw: t.is_nsfw })),
    purity: item.purity || 'sfw',
    category: item.category || 'general',
    file_size: item.file_size || 0,
    upload_date: item.created_at || '',
    likes: item.favorites || 0,
    views: item.views || 0,
  }))

  return { items, meta }
}

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(0)} KB`
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`
}

export async function fetchImageAsBase64(url: string): Promise<string> {
  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), 20000)
  const res = await fetch(url, {
    headers: { 'User-Agent': UA },
    signal: controller.signal,
  })
  clearTimeout(timeout)
  if (!res.ok) throw new Error(`fetch failed: ${res.status}`)
  const buf = await res.arrayBuffer()
  const base64 = Buffer.from(buf).toString('base64')
  const ext = url.split('.').pop()?.split('?')[0] || 'jpg'
  const mimes: Record<string, string> = { png: 'image/png', bmp: 'image/bmp', webp: 'image/webp', gif: 'image/gif' }
  const mime = mimes[ext] || 'image/jpeg'
  return `data:${mime};base64,${base64}`
}

export async function downloadToFile(url: string, destPath: string): Promise<string> {
  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), 60000)
  const res = await fetch(url, {
    headers: { 'User-Agent': UA },
    signal: controller.signal,
  })
  clearTimeout(timeout)
  if (!res.ok) throw new Error(`download failed: ${res.status}`)
  const buf = await res.arrayBuffer()
  const { writeFile } = await import('fs/promises')
  await writeFile(destPath, Buffer.from(buf))
  return destPath
}

export function extFromUrl(url: string): string {
  const match = url.match(/\.(\w+)(?:\?|#|$)/)
  return match ? match[1] : 'jpg'
}
